proposals/p2360.md

```diff
@@ -0,0 +1,291 @@
+        annotated in whatever way we eventually decide to annotate functions
+        that are usable at compile time, and values of that type will be usable
+        as types, in the same way that `()` and `{}` are.
+-   There are no singleton type-of-types. It's still possible to have, for
+    example, the type of `i32` be a type whose only value is `i32`. The value
+    `i32` would then not formally be a type, as its type would not be `Type`,
+    but it could still implicitly convert to `Type`, as `()` does, if we so
+    desire.
+    -   This extends to other cases: we could say that for integer literals,
+        which we have recently been describing as having type `IntLiteral(n)`,
+        there is an `impl IntLiteral(n) as ImplicitAs(Type)` which converts a
+        value of type `IntLiteral(n)` to the `Type` value `IntLiteral(n)`. This
+        would then permit `let v: 0 = 0;`.
+    -   Such changes to `i32` or integer literals are not part of this proposal.
+
+## Details
+
+### Terminology
+
+The changes to the language rules in this proposal are fairly small, but there
+is a significant change in how we describe those rules. We now use the following
+terminology:
+
+-   A _type_ is a value of type `Type`.
+-   A _kind_ is a constrained or unconstrained version of `Type`, such as an
+    interface type or a named constraint type. We previously referred to a kind
+    as a type-of-type, but that is no longer accurate, as the type of any type
+    is now by definition `Type`.
+    -   When one kind has a superset of the constraints of another kind, we say
+        it is a _subkind_. For example, `Hashable & Movable` is a subkind of
+        `Hashable`, which is a subkind of `Type`.
+-   A _facet_ is a value of a kind other than `Type`. For example,
+    `i32 as Hashable` is a facet. Note that facets are not types.
+-   We use the term _generic type_ to refer to a type or facet introduced by a
+    `:!` binding, such as in a generic parameter or associated constant.
+    -   This gives rise to terms such as _generic type parameter_ and
+        _associated generic type_.
+    -   It is worth noting that a generic type is not necessarily a type, as its
+        type may be an interface or constraint rather than `Type`.
+
+### Compound member access
+
+The syntax `x.(Y)`, where `Y` names an instance member, now always performs
+instance binding. Therefore, for a suitable `DebugPrintable`:
+
+-   `1.(DebugPrintable.Print)()` still prints `1`, as before.
+-   `i32.(DebugPrintable.Print)()` now prints `i32`, rather than forming a
+    member name.
+-   `1.(i32.(DebugPrintable.Print))()` is now an error, rather than printing
+    `1`.
+
+In order to get the old effect of `x.(MyType.(MyInterface.InstanceMember))()`,
+one can now write `x.((MyType as MyInterface).InstanceMember)()`. This behaves
+the same as the member access in:
+
+```
+fn F[MyType:! MyInterface, X:! Type](x: X) -> auto {
+  return x.(MyType.InstanceMember)();
```

Comment:
This example is doesn't look legal, since there is no way to deduce `MyType`. The smallest delta to make it compile would be:
```suggestion
fn F[MyType:! MyInterface, X:! MyType](x: X) -> auto {
  return x.(MyType.InstanceMember)();
```
But that doesn't really make sense since it defines `x.((MyType as MyInterface).InstanceMember)()` as `F(x)` which has free references to `MyInterface` and `InstanceMember` but not `MyType`.

I think the signature should instead be:
```suggestion
fn F(MyType:! MyInterface, x: MyType) -> auto {
  return x.(MyType.InstanceMember)();
```
And now `x.((MyType as MyInterface).InstanceMember)()` is equivalent to `F(MyType, x);`